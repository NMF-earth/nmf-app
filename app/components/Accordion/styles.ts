import { StyleSheet } from "react-native";

import { Colors } from "style";

export default StyleSheet.create({
  container: {
    flex: 1,
  },
});

export const itemStyles = StyleSheet.create({
  container: {
    backgroundColor: Colors.white,
    borderColor: Colors.primary10,
    borderRadius: 10,
    borderWidth: 1.5,
    marginTop: 16,
  },
  firstContainer: {
    marginTop: 0,
  },
  icon: {
    color: Colors.blue50,
  },
  titleContainer: {
    padding: 16,
    flexDirection: "row",
    alignItems: "center",
  },
  titleText: {
    color: Colors.black,
    fontSize: 16,
    marginLeft: 8,
    flex: 1,
  },
  descriptionContainer: {
    borderTopColor: Colors.primary10,
    borderTopWidth: 1.5,
    padding: 16,
  },
  descriptionText: {
    color: Colors.black70,
    fontSize: 12,
  },
});
