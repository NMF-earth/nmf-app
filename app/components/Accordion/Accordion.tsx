import React, { FC, ReactNode, Children, cloneElement, useState, ReactElement } from "react";
import { Pressable, View } from "react-native";
import { Ionicons } from "@expo/vector-icons";

import Text from "../Text";
import styles, { itemStyles } from "./styles";

interface ItemProps {
  title: string;
  children?: ReactNode;
  isFirst?: boolean;
  isOpen?: boolean;
  onTitlePress?: () => void;
}

interface AccordionProps {
  children?: ReactNode;
}

const Item: FC<ItemProps> = ({ title, children, isFirst, isOpen, onTitlePress }) => (
  <View style={[itemStyles.container, isFirst && itemStyles.firstContainer]}>
    <Pressable style={itemStyles.titleContainer} onPress={onTitlePress}>
      <Ionicons name={isOpen ? "chevron-down" : "chevron-up"} size={22} style={itemStyles.icon} />

      <Text.H3 style={itemStyles.titleText}>{title}</Text.H3>
    </Pressable>

    {isOpen ? (
      <View style={itemStyles.descriptionContainer}>
        <Text.Primary style={itemStyles.descriptionText}>{children}</Text.Primary>
      </View>
    ) : null}
  </View>
);

const Accordion: FC<AccordionProps> & { Item: typeof Item } = ({ children }) => {
  const [openIndex, setOpenIndex] = useState<number | undefined>();

  const toggleItem = (index: number) => {
    setOpenIndex((prev) => {
      if (prev === index) {
        return;
      }

      return index;
    });
  };

  return (
    <View style={styles.container}>
      {Children.map(children, (child, index) =>
        cloneElement(child as ReactElement, {
          isFirst: index === 0,
          isOpen: index === openIndex,
          onTitlePress: () => {
            toggleItem(index);
          },
        })
      )}
    </View>
  );
};

Accordion.Item = Item;

export default Accordion;
export { Item };
