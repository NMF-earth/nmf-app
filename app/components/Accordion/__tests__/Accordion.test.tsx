import React from "react";
import { create } from "react-test-renderer";

import Accordion from "../";

jest.unmock("../");

describe("Accordion", () => {
  it("renders correctly", () => {
    const tree = create(
      <Accordion>
        <Accordion.Item title="Test title 1">Test Description 1</Accordion.Item>

        <Accordion.Item title="Test title 2">Test Description 2</Accordion.Item>
      </Accordion>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
