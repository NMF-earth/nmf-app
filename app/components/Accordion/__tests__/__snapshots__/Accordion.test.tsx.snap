// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#E7F3EB",
          "borderRadius": 10,
          "borderWidth": 1.5,
          "marginTop": 16,
        },
        {
          "marginTop": 0,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 16,
        }
      }
    >
      <Ionicons
        name="chevron-up"
        size={22}
        style={
          {
            "color": "#3957D0",
          }
        }
      />
      <Text.H3
        style={
          {
            "color": "#0F0F0F",
            "flex": 1,
            "fontSize": 16,
            "marginLeft": 8,
          }
        }
      >
        Test title 1
      </Text.H3>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#E7F3EB",
          "borderRadius": 10,
          "borderWidth": 1.5,
          "marginTop": 16,
        },
        false,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 16,
        }
      }
    >
      <Ionicons
        name="chevron-up"
        size={22}
        style={
          {
            "color": "#3957D0",
          }
        }
      />
      <Text.H3
        style={
          {
            "color": "#0F0F0F",
            "flex": 1,
            "fontSize": 16,
            "marginLeft": 8,
          }
        }
      >
        Test title 2
      </Text.H3>
    </View>
  </View>
</View>
`;
